File: sources/android/app/src/main/java/com/siarheikuchuk/ftpsserver/privacy/PrivacyLanguages.kt
package com.siarheikuchuk.ftpsserver.privacy

import java.util.Locale

data class PrivacyLanguage(
    val code: String,
    val assetFile: String,
    val nativeName: String,
    val rtl: Boolean = false,
)

object PrivacyLanguages {
    val all: List<PrivacyLanguage> = listOf(
        PrivacyLanguage("en", "en.md", "English"),
        PrivacyLanguage("am", "am.md", "አማርኛ"),
        PrivacyLanguage("ar", "ar.md", "العربية", rtl = true),
        PrivacyLanguage("bn", "bn.md", "বাংলা"),
        PrivacyLanguage("my", "my.md", "မြန်မာ"),
        PrivacyLanguage("yue", "yue.md", "粵語"),
        PrivacyLanguage("zh-Hans", "zh-Hans.md", "简体中文"),
        PrivacyLanguage("fr", "fr.md", "Français"),
        PrivacyLanguage("de", "de.md", "Deutsch"),
        PrivacyLanguage("ha", "ha.md", "Hausa"),
        PrivacyLanguage("hi", "hi.md", "हिन्दी"),
        PrivacyLanguage("ig", "ig.md", "Igbo"),
        PrivacyLanguage("id", "id.md", "Bahasa Indonesia"),
        PrivacyLanguage("it", "it.md", "Italiano"),
        PrivacyLanguage("ja", "ja.md", "日本語"),
        PrivacyLanguage("kk", "kk.md", "Қазақша"),
        PrivacyLanguage("ko", "ko.md", "한국어"),
        PrivacyLanguage("mr", "mr.md", "मराठी"),
        PrivacyLanguage("ne", "ne.md", "नेपाली"),
        PrivacyLanguage("pcm", "pcm.md", "Nigerian Pidgin"),
        PrivacyLanguage("om", "om.md", "Afaan Oromoo"),
        PrivacyLanguage("ps", "ps.md", "پښتو", rtl = true),
        PrivacyLanguage("fa", "fa.md", "فارسی", rtl = true),
        PrivacyLanguage("pl", "pl.md", "Polski"),
        PrivacyLanguage("pt-BR", "pt-BR.md", "Português (Brasil)"),
        PrivacyLanguage("pa", "pa.md", "ਪੰਜਾਬੀ"),
        PrivacyLanguage("ru", "ru.md", "Русский"),
        PrivacyLanguage("es", "es.md", "Español"),
        PrivacyLanguage("sw", "sw.md", "Kiswahili"),
        PrivacyLanguage("ta", "ta.md", "தமிழ்"),
        PrivacyLanguage("te", "te.md", "తెలుగు"),
        PrivacyLanguage("th", "th.md", "ไทย"),
        PrivacyLanguage("tr", "tr.md", "Türkçe"),
        PrivacyLanguage("uk", "uk.md", "Українська"),
        PrivacyLanguage("ur", "ur.md", "اردو", rtl = true),
        PrivacyLanguage("uz", "uz.md", "Oʻzbekcha"),
        PrivacyLanguage("vi", "vi.md", "Tiếng Việt"),
        PrivacyLanguage("yo", "yo.md", "Yorùbá"),
    )

    fun byCode(code: String): PrivacyLanguage =
        all.firstOrNull { it.code.equals(code, ignoreCase = true) } ?: all.first { it.code == "en" }

    fun matchDevice(locale: Locale = Locale.getDefault()): PrivacyLanguage {
        val lang = locale.language.lowercase(Locale.ROOT)
        val country = locale.country.uppercase(Locale.ROOT)
        val script = locale.script
        when {
            lang == "zh" && (script.equals("Hant", true) || country in setOf("HK", "MO", "TW")) ->
                return byCode("yue")
            lang == "zh" -> return byCode("zh-Hans")
            lang == "pt" -> return byCode("pt-BR")
            lang == "in" || lang == "id" -> return byCode("id")
            lang == "yue" -> return byCode("yue")
            lang == "pcm" -> return byCode("pcm")
        }
        return all.firstOrNull { it.code.equals(lang, ignoreCase = true) } ?: byCode("en")
    }
}

File: sources/FtpsServerAppsShared/Privacy/PrivacyLanguages.cs
using System.Globalization;

namespace FtpsServerAppsShared.Privacy;

public static class PrivacyLanguages
{
    public static IReadOnlyList<PrivacyLanguage> All { get; } =
    [
        new("en", "en.md", "English"),
        new("am", "am.md", "አማርኛ"),
        new("ar", "ar.md", "العربية", Rtl: true),
        new("bn", "bn.md", "বাংলা"),
        new("my", "my.md", "မြန်မာ"),
        new("yue", "yue.md", "粵語"),
        new("zh-Hans", "zh-Hans.md", "简体中文"),
        new("fr", "fr.md", "Français"),
        new("de", "de.md", "Deutsch"),
        new("ha", "ha.md", "Hausa"),
        new("hi", "hi.md", "हिन्दी"),
        new("ig", "ig.md", "Igbo"),
        new("id", "id.md", "Bahasa Indonesia"),
        new("it", "it.md", "Italiano"),
        new("ja", "ja.md", "日本語"),
        new("kk", "kk.md", "Қазақша"),
        new("ko", "ko.md", "한국어"),
        new("mr", "mr.md", "मराठी"),
        new("ne", "ne.md", "नेपाली"),
        new("pcm", "pcm.md", "Nigerian Pidgin"),
        new("om", "om.md", "Afaan Oromoo"),
        new("ps", "ps.md", "پښتو", Rtl: true),
        new("fa", "fa.md", "فارسی", Rtl: true),
        new("pl", "pl.md", "Polski"),
        new("pt-BR", "pt-BR.md", "Português (Brasil)"),
        new("pa", "pa.md", "ਪੰਜਾਬੀ"),
        new("ru", "ru.md", "Русский"),
        new("es", "es.md", "Español"),
        new("sw", "sw.md", "Kiswahili"),
        new("ta", "ta.md", "தமிழ்"),
        new("te", "te.md", "తెలుగు"),
        new("th", "th.md", "ไทย"),
        new("tr", "tr.md", "Türkçe"),
        new("uk", "uk.md", "Українська"),
        new("ur", "ur.md", "اردو", Rtl: true),
        new("uz", "uz.md", "Oʻzbekcha"),
        new("vi", "vi.md", "Tiếng Việt"),
        new("yo", "yo.md", "Yorùbá"),
    ];

    public static PrivacyLanguage ByCode(string? code) =>
        All.FirstOrDefault(l => string.Equals(l.Code, code, StringComparison.OrdinalIgnoreCase))
        ?? All.First(l => l.Code == "en");

    public static PrivacyLanguage MatchDevice(CultureInfo? culture = null)
    {
        culture ??= CultureInfo.CurrentUICulture;
        var lang = culture.TwoLetterISOLanguageName.ToLowerInvariant();
        var country = culture.Name.Contains('-')
            ? culture.Name[(culture.Name.IndexOf('-') + 1)..].ToUpperInvariant()
            : culture.Name.ToUpperInvariant();
        var script = culture.Name;

        if (lang == "zh" && (script.Contains("Hant", StringComparison.OrdinalIgnoreCase)
            || country is "HK" or "MO" or "TW" or "HANT"))
            return ByCode("yue");
        if (lang == "zh")
            return ByCode("zh-Hans");
        if (lang == "pt")
            return ByCode("pt-BR");
        if (lang is "in" or "id")
            return ByCode("id");
        if (lang == "yue")
            return ByCode("yue");
        if (lang == "pcm")
            return ByCode("pcm");

        return All.FirstOrDefault(l => l.Code.Equals(lang, StringComparison.OrdinalIgnoreCase))
            ?? ByCode("en");
    }
}
